packages/shared/src/hooks/translation/useTranslation.ts

```diff
@@ -0,0 +1,141 @@
+import { useCallback, useEffect, useRef } from 'react';
+import type { InfiniteData, QueryKey } from '@tanstack/react-query';
+import { useQueryClient } from '@tanstack/react-query';
+import { stream } from 'fetch-event-stream';
+import { useAuthContext } from '../../contexts/AuthContext';
+import { apiUrl } from '../../lib/config';
+import type { FeedData, Post } from '../../graphql/posts';
+import {
+  updateCachedPagePost,
+  findIndexOfPostInData,
+  updatePostCache,
+} from '../../lib/query';
+
+export enum ServerEvents {
+  Connect = 'connect',
+  Message = 'message',
+  Disconnect = 'disconnect',
+  Error = 'error',
+}
+
+type UseTranslation = (props: {
+  queryKey: QueryKey;
+  queryType: 'post' | 'feed';
+}) => {
+  fetchTranslations: (id: string[]) => void;
+};
+
+type TranslateEvent = {
+  id: string;
+  title: string;
+};
+
+const updateTranslation = (post: Post, translation: TranslateEvent): Post => {
+  const updatedPost = post;
+  if (post.title) {
+    updatedPost.title = translation.title;
+    updatedPost.translation = { title: !!translation.title };
+  } else {
+    updatedPost.sharedPost.title = translation.title;
+    updatedPost.sharedPost.translation = { title: !!translation.title };
+  }
+
+  return updatedPost;
+};
+
+export const useTranslation: UseTranslation = ({ queryKey, queryType }) => {
+  const abort = useRef<AbortController>();
+  const { user, accessToken, isLoggedIn } = useAuthContext();
+  const queryClient = useQueryClient();
+
+  const { language } = user || {};
+
+  const updateFeed = useCallback(
+    (translatedPost: TranslateEvent) => {
+      const updatePost = updateCachedPagePost(queryKey, queryClient);
+      const feedData =
+        queryClient.getQueryData<InfiniteData<FeedData>>(queryKey);
+      const { pageIndex, index } = findIndexOfPostInData(
+        feedData,
+        translatedPost.id,
+        true,
+      );
+      if (index > -1) {
+        updatePost(
+          pageIndex,
+          index,
+          updateTranslation(
+            feedData.pages[pageIndex].page.edges[index].node,
+            translatedPost,
+          ),
+        );
+      }
+    },
+    [queryKey, queryClient],
+  );
+
+  const updatePost = useCallback(
+    (translatedPost: TranslateEvent) => {
+      updatePostCache(queryClient, translatedPost.id, (post) =>
+        updateTranslation(post, translatedPost),
+      );
+    },
+    [queryClient],
+  );
+
+  const fetchTranslations = useCallback(
+    async (postIds: string[]) => {
+      if (!isLoggedIn || !language) {
+        return;
+      }
```

Comment:
Merge it to `isStreamActive` and only provide single depth to `useEffect`?

Reply:
Check 020414f